src/common-ui/containers/AddListDropdownContainer.js

```diff
@@ -0,0 +1,358 @@
+        applyBulkEdits: PropTypes.func,
+        resetPagesinTempList: PropTypes.func,
+        setTempLists: PropTypes.func,
+        mode: PropTypes.string.isRequired,
+        url: PropTypes.string,
+    }
+
+    static defaultProps = {
+        onFilterAdd: noop,
+        onFilterDel: noop,
+        results: [],
+        displayFilters: [],
+    }
+
+    constructor(props) {
+        super(props)
+
+        this.fetchListSuggestions = debounce(300)(this.fetchListSuggestions)
+        this.addList = remoteFunction('createCustomList')
+        this.addUrlToList = remoteFunction('insertPageToList')
+        this.fetchListNameSuggestions = remoteFunction(
+            'fetchListNameSuggestions',
+        )
+
+        this.state = {
+            searchVal: '',
+            isLoading: false,
+            displayFilters: props.initSuggestions
+                ? props.initSuggestions
+                : props.results, // Display state objects; will change all the time
+            filters: props.results, // Actual lists associated with the page; will only change when DB updates
+            focused: props.results.length ? 0 : -1,
+        }
+    }
+
+    componentWillMount() {
+        // The temporary list array gets updated.
+        if (this.overviewMode) this.props.setTempLists()
+    }
+
+    get inputBlockPattern() {
+        return /[^\w\s-]/gi
+    }
+
+    get allowIndexUpdate() {
+        return this.props.url != null
+    }
+
+    get overviewMode() {
+        return this.props.mode === 'overview'
+    }
+
+    scrollElementIntoViewIfNeeded(domNode) {
+        const parentNode = domNode.parentNode
+        parentNode.scrollTop = domNode.offsetTop - parentNode.offsetTop
+    }
+
+    setInputRef = el => (this.inputEl = el)
+
+    canCreateList() {
+        if (!this.allowIndexUpdate) {
+            return false
+        }
+
+        const searchVal = this.getSearchVal()
+
+        return (
+            !!searchVal.length &&
+            !this.state.displayFilters.reduce(
+                (acc, tag) => acc || tag.name === searchVal,
+                false,
+            )
+        )
+    }
+
+    addPageToList = () => { }
+
+    createList = async () => {
+        let newLists = this.state.filters
+
+        try {
+            if (this.allowIndexUpdate) {
+                // Adds a list as well as add this page to list.
+                // const id = await this.addList({ name: this.getSearchVal() })
+                this.addList({ name: this.getSearchVal() }).then((id) => {
+                    const newList = {
+                        id,
+                        name: this.getSearchVal(),
+                        pages: [],
+                        active: true,
+                    }
+                    this.addUrlToList({
+                        id,
+                        url: this.props.url,
+                    })
+                    newLists = [newList, ...this.state.filters]
+                    this.setState(state => ({
```

Comment:
As far as I know, setState() only overwrites the properties you give to it.

Reply:
According to the docs, it seems like it. TIL (http://memex.link/eAC07tnVq/reactjs.org/docs/react-component.html#setstate) I have always manually merged the prev state, I guess as a holdover from redux reducers.

I somehow skipped over this part of the changes today, but looking now I don't think there is any need for the `try...catch...finally` anymore, right? You should also have a `.catch()` on the `this.addUrlToList` call that should revert back the optimistic state changes (like the opposite of the current `setState` call) in-case anything goes bad.